apiserver/objects_test.go

```diff
@@ -219,11 +326,276 @@ func (s *putCharmObjectSuite) TestUploadBumpsRevision(c *gc.C) {
 	c.Assert(sch.BundleSha256(), gc.Not(gc.Equals), "")
 }
 
+func (s *putCharmObjectSuite) TestUploadVersion(c *gc.C) {
+	expectedVersion := "dummy-146-g725cfd3-dirty"
+
+	// Add the dummy charm with version "juju-2.4-beta3-146-g725cfd3-dirty".
+	pathToArchive := testcharms.Repo.CharmArchivePath(c.MkDir(), "dummy")
+	err := testcharms.InjectFilesToCharmArchive(pathToArchive, map[string]string{
+		"version": expectedVersion,
+	})
+	c.Assert(err, gc.IsNil)
+	ch, err := charm.ReadCharmArchive(pathToArchive)
+	c.Assert(err, gc.IsNil)
+
+	f, err := os.Open(ch.Path)
+	c.Assert(err, jc.ErrorIsNil)
+	defer f.Close()
+	resp := s.uploadRequest(c, s.objectsCharmsURI("dummy-"+getCharmHash(c, f)), "application/zip", "local:dummy", f)
+
+	expectedURL := "local:dummy-1"
+	s.assertUploadResponse(c, resp, expectedURL)
+	sch, err := s.ControllerModel(c).State().Charm(expectedURL)
+	c.Assert(err, jc.ErrorIsNil)
+
+	version := sch.Version()
+	c.Assert(version, gc.Equals, expectedVersion)
+}
+
+func (s *putCharmObjectSuite) TestUploadRespectsLocalRevision(c *gc.C) {
+	// Make a dummy charm dir with revision 123.
+	dir := testcharms.Repo.ClonedDir(c.MkDir(), "dummy")
+	dir.SetDiskRevision(123)
+	// Now bundle the dir.
```

Comment:
"bundle" in this context is a very old concept. It's confusing enough in the uniter code. Please update to "tar" or "archive" rather than an overloaded term like "bundle"